### DISCUSSION TITLE
Testing with a parameterized AuthManagerLayer

### DISCUSSION THREAD

**Discussion by capveg-netdebug** [maxcountryman/axum-login#192] [Q&A]
Hi all,

Thanks for the nice crate.  I was trying to create a "create all my routes" function that took an `AuthManagerLayer` as a parameter so I could pass in different settings (e.g., session store, authentication mechanisms, etc.) for testing verses production.   I've run into two problem: one which I've hacked around (I think) and the other I'm stuck on and I'm wondering what others do here.  I've looked at the tests in https://docs.rs/axum-login/latest/src/axum_login/middleware.rs.html#348 but they all seem to take a literal AuthManagerSession which dodges these problems.  See the code below for what I'm trying to do.

Problem 1: the `login_required!()` wants a literal type and thus can't take a generic type.  I think this might be fixable by changing the ":ty" in https://github.com/maxcountryman/axum-login/blob/main/axum-login/src/middleware.rs#L41 to an ":expr" but I've been able to hack around this by manually expanding the macro.  Thoughts?

Problem 2: while this code works for me in production and works in some of my tests (e.g., calls to "/test_auth"; first test below) other tests (e.g. the second testauth_check_with_auth() below) fails out with an internal "Can't extract auth session. Is `AuthManagerLayer` enabled?" -- where I'm definitely calling the `.layer(auth_layer)` call.

I feel like I'm abusing the APIs and tripping over internal assumptions... any thoughts on how to do this successfully?  Happy to throw up some fixes if there's agreement that something needs to be fixed/changed.

Thanks in advance,

- Rob
.

```rust
pub fn setup_protected_rest_routes_with_auth_layer<
    Backend: AuthnBackend + Clone + 'static,
    Session: SessionStore + Clone,
>(
    auth_layer: AuthManagerLayer<Backend, Session>,
) -> Router<Context> {
    // Manually expanded helper from login_required!(); NOTE 'Backend2' is the same type as 'Backend'
    // but rustc doesn't want to inherit the definition from the outer scope
    async fn is_authenticated<Backend2: AuthnBackend + Clone + 'static>(
        auth_session: axum_login::AuthSession<Backend2>,
    ) -> bool {
        auth_session.user.is_some()
    }

    Router::new()
        // list the paths that need authentication here
        .route("/test_auth", routing::get(test_auth))
        // don't use the login_required!() macro : can't figure out generic types so manually expand
        // .router_layer(login_required!(Backend)) --> causes compiler error!
        .route_layer(predicate_required!(
            is_authenticated::<Backend>,
            StatusCode::UNAUTHORIZED
        ))
        // these are unauthenticated routes to get the auth token
        .route("/login", routing::post(console_login))
        .route("/login", routing::get(console_login))
        .layer(auth_layer)
}

// simple protected function to make sure that auth is setup
async fn test_auth(auth_session: AuthSession) -> impl IntoResponse {
    match auth_session.user {
        Some(user) => (StatusCode::OK, format!("Hello {}", user.id())),
        None => (
            StatusCode::INTERNAL_SERVER_ERROR,
            "Weird - authenticated user not found!?".to_string(),
        ),
    }
    .into_response()
}
 
 // simple function to issue a session token
 async fn console_login(
    mut auth_session: AuthSession,
    Form(creds): Form<AuthCredentials>,
) -> impl IntoResponse {
    let user = match auth_session.authenticate(creds.clone()).await {
        Ok(Some(user)) => user,
        Ok(None) => return StatusCode::UNAUTHORIZED.into_response(),
        Err(e) => return (StatusCode::INTERNAL_SERVER_ERROR, format!("{}", e)).into_response(),
    };

    if auth_session.login(&user).await.is_err() {
        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
    }

    StatusCode::OK.into_response()
}
 
    async fn make_mock_protected_routes() -> Router {
        let mut mock_backend = MockAuthBackend::default();
        let user = "alice".to_string();
        mock_backend.add_user(
            user.clone(),
            User {
                user_id: user.clone(),
                session_key: User::make_session_key(&user, &"random".to_string()),
            },
        );
        let context = make_test_context();
        let session_store = MemoryStore::default();
        let session_layer = SessionManagerLayer::new(session_store).with_secure(false);
        let auth_layer = AuthManagerLayerBuilder::new(mock_backend, session_layer).build();
        setup_protected_rest_routes_with_auth_layer(auth_layer).with_state(context)
        // context not needed except to make compiler happy
    }

    #[tokio::test]
    async fn auth_check_no_auth() {
        let protected_routes = make_mock_protected_routes().await;

        // `Router` implements `tower::Service<Request<Body>>` so we can
        // call it like any tower service, no need to run an HTTP server.
        let response = protected_routes
            .oneshot(
                Request::builder()
                    .uri("/test_auth")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        let status = response.status();
        // let body = response.into_body().collect().await.unwrap().to_bytes();
        // println!("Response={:#?}", body);

        assert_eq!(status, StatusCode::UNAUTHORIZED);
    }
    #[tokio::test]
    async fn auth_check_with_auth() {
        let protected_routes = make_mock_protected_routes().await;

        // `Router` implements `tower::Service<Request<Body>>` so we can
        // call it like any tower service, no need to run an HTTP server.
        let response = protected_routes
            .oneshot(
                Request::builder()
                    .uri("/login")
                    .body(Body::from(
                        serde_json::to_string(&AuthCredentials {
                            jwt: "Alice".to_string(),
                        })
                        .unwrap(),
                    ))
                    .unwrap(),
            )
            .await
            .unwrap();

        let status = response.status();
        let cookie = get_session_cookie(&response);
        let body = response.into_body().collect().await.unwrap().to_bytes();
        println!("Response={:#?}", body);
        println!("Cookie={:?}", cookie);
        assert_eq!(status, StatusCode::OK);
    }
  
    ```

**Comment by maxcountryman** [comment]
> Problem 1: the login_required!() wants a literal type and thus can't take a generic type. I think this might be fixable by changing the ":ty"

I believe that won't work as we'd hope--without a type literal, the macro doesn't compile. That said, if I missed something or there's other ways we might improve the ergonomics I think it's worth a look.

As far as your second problem, have you considered writing your own macro?

I don't love macros because they're generally difficult to compose, but when it comes to producing this middleware there seemingly aren't many great options. Therefore we've very much left the door open to library users implement there own middleware here via `from_fn`, etc.

**Reply by capveg-netdebug** [reply]
Thank you for the reply... let me look into the macro option.  I had assumed (hoped?) there would be a non-macro solution to this but if that's your best suggestion, I appreciate the help.  Thank you!

**Reply by maxcountryman** [reply]
For testing especially I often reach for macros.

That said, if there are ways we can improve the crate API to make this more ergonomic for you I'm happy to explore what that could look like.

**Reply by capveg-netdebug** [reply]
Thanks - any insight why I'm getting the "Can't extract auth session. Is AuthManagerLayer enabled?" error in my test?  That's the piece I'd like to understand before I consider trying to suggest an improvement.  Thank you!

**Reply by maxcountryman** [reply]
The error message tries to explain the problem: the AuthManagerLayer isn't installed such that a session can be extracted. Usually this means you either forgot to install the layer altogether or you're using incompatible versions of the session (the layer was setup with a session version that differs from the one your code is actually using).
